crates/containerd-shim-wasm/src/sandbox/sync.rs

```diff
@@ -0,0 +1,117 @@
+// this is safe because access to cell guarded by the mutex
+unsafe impl<T> Send for WaitableCell<T> {}
+unsafe impl<T> Sync for WaitableCell<T> {}
+
+struct WaitableCellSetGuard<T, R: Into<T>, F: FnOnce() -> R> {
+    f: Option<F>,
+    cell: WaitableCell<T>,
+}
+
+impl<T> Default for WaitableCell<T> {
+    fn default() -> Self {
+        Self {
+            inner: Arc::new(WaitableCellImpl {
+                mutex: Mutex::new(()),
+                cvar: Condvar::new(),
+                cell: OnceCell::new(),
+            }),
+        }
+    }
+}
+
+impl<T> Clone for WaitableCell<T> {
+    fn clone(&self) -> Self {
+        let inner = self.inner.clone();
+        Self { inner }
+    }
+}
+
+impl<T> WaitableCell<T> {
+    /// Creates an empty WaitableCell.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Creates a WaitableCell with a value.
+    pub fn with_value(val: T) -> Self {
+        let cell = Self::default();
+        let _ = cell.set(val);
+        cell
+    }
+
+    /// Sets a value to the WaitableCell.
+    /// This method has no effect if the WaitableCell already has a value.
+    pub fn set(&self, val: impl Into<T>) -> Result<(), T> {
+        let val = val.into();
+        let _guard = self.inner.mutex.lock().unwrap();
+        let res = self.inner.cell.set(val);
+        self.inner.cvar.notify_all();
+        res
+    }
+
+    /// If the `WaitableCell` is empty when this guard is dropped, it will set with `val`.
+    pub fn set_guard(&self, val: impl Into<T>) -> impl Drop {
+        self.set_guard_with(move || val.into())
+    }
+
+    /// If the `WaitableCell` is empty when this guard is dropped, it will set with the result of `f`.
+    pub fn set_guard_with<R: Into<T>>(&self, f: impl FnOnce() -> R) -> impl Drop {
+        let cell = (*self).clone();
+        WaitableCellSetGuard { f: Some(f), cell }
+    }
+
+    /// Wait for the WaitableCell to be set a value.
+    pub fn wait(&self) -> &T {
+        let value = self.wait_timeout(None);
+        // safe because we waited with timeout `None`
+        unsafe { value.unwrap_unchecked() }
+    }
+
+    /// Wait for the WaitableCell to be set a value, with timeout.
+    /// Retuns None if the timeout is reached with no value.
+    pub fn wait_timeout(&self, timeout: impl Into<Option<Duration>>) -> Option<&T> {
+        let timeout = timeout.into();
+        let cvar = &self.inner.cvar;
+        let guard = self.inner.mutex.lock().unwrap();
+        let _guard = match timeout {
+            None => cvar
+                .wait_while(guard, |_| self.inner.cell.get().is_none())
+                .unwrap(),
+            Some(Duration::ZERO) => guard,
+            Some(dur) => cvar
+                .wait_timeout_while(guard, dur, |_| self.inner.cell.get().is_none())
+                .map(|(guard, _)| guard)
+                .unwrap(),
+        };
+        self.inner.cell.get()
+    }
+}
+
+impl<T, R: Into<T>, F: FnOnce() -> R> Drop for WaitableCellSetGuard<T, R, F> {
```

Comment:
Could you please add some comments explaining this usage?

Reply:
This type is internal to the module. It is an implementation details on how `set_guard_with` works.
I've added a few comments on this type, and more detaisl comments in `set_guard_with`.